### PR TITLE
feat: added pydantic models for Python

### PR DESCRIPTION
**Pull request by camilovelezr**
This PR:

- Creates a Python API
- Defines a Pydantic schema model based on [this spec](https://github.com/foundationopenresearch/ict-spec/blob/main/spec.md)

Examples:
```python3
import ict
my_example_ict_spec =  {
        "specVersion": "0.1.2",
        "name": "test/t2",
        "version": "0.1.0",
        "container": "wipp/wipp-backend:latest",
        "entrypoint": "test.py",
        "title": "Test Title",
        "description": "Test Description",
        "author": ["John Doe", "Jane Doe"],
        "contact": "c@u.edu",
        "repository": "https://github.com/ict",
        "citation": "10.5281/zenodo.1234",
        "inputs": [
            {
                "name": "foo",
                "type": "string",
                "description": "bar",
                "required": True,
                "format": ["es"],
            }
        ],
        "outputs": [
            {
                "name": "bar",
                "type": "array",
                "description": "bar",
                "required": "no",
                "format": ["es"],
            }
        ],
        "ui": [
            {
                "key": "inputs.foo",
                "title": "foo",
                "description": "bar",
                "type": "number",
                "condition": "inputs.foo==1",
            },
            {
                "key": "inputs.bar",
                "title": "foo",
                "description": "bar",
                "type": "text",
            },
        ],
    }
ict.validate(my_example_ict_spec)

ict.validate("/home/ict-user/my_awesome_ict.yml")
```

Running `ict.validate()` will perform validation and, if there are no errors, it will return an instance of `ICT`

### CODE REVIEW

**Comment by agerardin** [comment]
Overall the code is very readable and well organized.


One remark :

- your example references spec 1.2, while the spec.md mention 1.0.
So it is not clear what is 1.2? The example you provide seems to vary slightly from the one in the repo. (ex for format, it is containing a single uri and term in 0.1.0 and it seems we have only list of string in 0.1.2 - I would guess we need a list of uri and term object).
- More generally, we probably need a versioning strategy for pydantic model because once the spec is in the wild, people will have different versions floating around.

One feature that may be needed :

- The tool seems to consume yaml. At least in compute we may also need to handle json but you may need to double-check that with the compute team.

The rest are comments for maintainability:

- Generally, i would recommend at least a few basic unit tests.
* At the api level, demonstrating that at least the canonical example for a version is run.
* as needed, for specific validation. that's really up to you but we make it easier for a maintainer to figure out the original intent, specially for regex
     ex:  the regex in check_container. My understanding is that no "_" are  allowed in registry path?

- If some parts of the model are coming from authoritative sources, we should provide a link.
ex : semver or w3date may or may not be, but that helps clarifying for the maintainer and maybe alleviate the need to write some more unit tests.

Those are all "extra-mile" considerations so feel free push to final review if needed.

File: src/ict/hardware/objects.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# pylint: disable=no-member
+"""Hardware Requirements for ICT."""
+from typing import Annotated, Optional, Union
+
+from pydantic import BaseModel, BeforeValidator, Field
+
+
+def validate_str(s_t: Optional[int]) -> Union[str, None]:
+    """Return a string from int."""
```
Comment (Nicholas-Schaub) [inline src/ict/hardware/objects.py]: If this is only supposed to be for `int`, we should add a check to make sure the input is an `int`.

If this should support `float`, we should update the type hint.

File: src/ict/io/objects.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    required: bool = Field(
+        description="Boolean (true/false) value indicating whether this "
+        + "field needs an associated value.",
+        examples=["true"],
+    )
+    io_format: Union[list[str], dict] = Field(
+        ...,
+        alias="format",
+        description="Defines the actual value(s) that the input/output parameter represents"
+        + "represents using an ontology schema.",
```
Comment (Nicholas-Schaub) [inline src/ict/io/objects.py:45]: Typo: repeated `represents`

File: src/ict/metadata/objects.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    root: str
+
+    @field_validator("root")
+    @classmethod
+    def check_author(cls, value):
+        """Check the author follows the correct format."""
+        if not len(value.split(" ")) == 2:
+            raise ValueError(
+                "The author must be in the format <first name> <last name>"
+            )
```
Comment (Nicholas-Schaub) [inline src/ict/metadata/objects.py:34]: Could this be problematic for people of  non-western cultures?

**Review by Nicholas-Schaub — CHANGES_REQUESTED** [review]
Just a couple of minor changes, and this looks good to go. I am very impressed with it.

File: src/ict/metadata/objects.py (hunk 3 above)
Comment (camilovelezr) [reply src/ict/metadata/objects.py:34]: @sunnielyu @NHotaling

**Pull request merged**